proposals/p1356.md

```diff
@@ -0,0 +1,376 @@
+for `Subscriptable`:
+
+```
+final external impl [IndexType:! Type, T:! SubscriptableAsRValue(IndexType)]
+    T as Subscriptable(IndexType) {
+  let ElementType:! Type = T.(SubscriptableAsRValue(IndexType)).ElementType;
+  fn Subscript[me: Self](index: IndexType) -> ElementType {
+    return *(me.SubscriptAsRValue(index));
+  }
+  fn SubscriptAsLValue[addr me: Self*](index: IndexType) -> ElementType* {
+    return me->SubscriptAsRValue(index);
+  }
+}
+```
+
+Thus, a type that implements `SubscriptableAsRValue` need not, and cannot,
+provide its own definitions of `Subscript` and `SubscriptAsLValue`.
+
+### Examples
+
+An array type could implement subscripting like so:
+
+```
+class Array(template T:! Type) {
+  external impl as Subscriptable(like i64) {
+    let ElementType:! Type = T;
+    fn Index[me: Self](subscript: i64) -> T;
+    fn LValueIndex[addr me: Self*](subscript: i64) -> T*;
+  }
+}
+```
+
+And a slice type could look like this:
+
+```
+class Slice(T:! Type) {
+  external impl as RValueSubscriptable(like i64) {
+    let ElementType:! Type = T;
+    fn RValueIndex[me: Self](subscript: i64) -> T*;
+  }
+}
+```
+
+We can even approximate support for indexing on tuple types. We don't have
+enough of a design for metaprogramming or variadics to specify it generically,
+but we can illustrate how it would look for a particular tuple type, such as
+`(i8, f64)`:
+
+```
+external impl (i8, f64) as Subscriptable(typeof(0)) {
+  let ElementType:! Type = i8;
+  fn Index[me: Self](0) -> i8;
+  fn LValueIndex[addr me: Self*](0) -> i8*;
+}
+
+external impl (i8, f64) as Subscriptable(typeof(1)) {
+  let ElementType:! Type = f64;
+  fn Index[me: Self](1) -> f64;
+  fn LValueIndex[addr me: Self*](1) -> f64*;
+}
+```
+
+However, this approach presupposes that every integer literal has a distinct
+type, which is not yet settled. It also only supports indexing with integer
+literals, not with constant values of ordinary integer types. This is somewhat
+contrary to Carbon's metaprogramming philosophy: Carbon treats types as values
+in order to make metaprogramming more like ordinary programming, but here we're
+doing the reverse, treating values as types. Consequently, the eventual design
+for tuple-style indexing may look very different.
+
+## Rationale based on Carbon's goals
+
+Indexing is a fundamental operation in procedural programming, and is especially
+common in performance-critical code, so supporting it with a concise, familiar
+syntax helps us make
+[code easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
+and support
+[performance-critical software](/docs/project/goals.md#performance-critical-software).
+Using the same syntax as C++, even for slice-like types, supports
+[interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code).
+
+## Alternatives considered
```

Comment:
I think it would be good to document the problems we found with selecting the interface to use based on whether the resulting value was used as an l-value or r-value.

Reply:
I think that's covered by the last paragraph of "map-like subscripting" below. Is there more you think I should cover there?